### ISSUE TITLE
How to Get Feature Pull Requests Accepted

### ISSUE THREAD

**Issue by gavinhoward** [memononen/nanovg#413]
I am currently working on a project that is using NanoVG. It's called [Wima](https://github.com/ProjectLFyre/Wima), and NanoVG is what does all of the drawing.

However, I have a problem: in order for Wima to be feature complete, I need a certain feature, specifically Picking/Hit Regions (#401). I have already created the pull request and I keep merging changes from this repo into mine to keep it merge-ready. Despite all of that, nothing has been said about it, even though bug fix pull requests have been accepted.

How can I get this pull request accepted?

**Comment by lieff** [comment]
Hmmm. I need this feature too, but I think about more general code instead on integrating it into nanovg. We only need function that tests set of path (described in form edge type + move\line\curve data) and returns result: point inside filled shape or not. I need this functionality across several backends, not only nanovg.

**Comment by fundamental** [comment]
Just out of curiosity, what do you need the hit/pick detection for in the context of your toolkit? Over the past few years I've build my own nanovg based toolkit and never found such functionality as a needed feature; Mouse handling was done via managing the widget (bounding box) geometry outside of nanovg and letting individual widgets manage clickable regions within their own bounding box via relatively simple geometry (porting the color picker from the example is an ongoing headache, but that's a rarity).

**Comment by gavinhoward** [comment]
@fundamental I am building a functional programming environment that is based on the idea of nodes linked together, much like [Blender](https://www.blender.org/)'s compositor.

![blender compositing](https://user-images.githubusercontent.com/3172687/30877872-511ab310-a2b8-11e7-8872-4211bfb78b43.jpg)

I need to be able to tell when the mouse intersects the links, and taking the bounding box isn't going to help very much, especially since the links can be made more or less curvy, depending on the user's preferences.

If you have another idea though, I would be very glad to hear it.

**Comment by fundamental** [comment]
Good answer. In the case of my toolkit the closest problem is the user selecting control points on a few (optionally curved) lines within an xy-plot widget. Since one of the targets of the interface is touch devices the touch region is defined by min_distance(line or target) < threshold (found algebraically or via calculated distance to a few sampled points) rather than if the line is selected itself. My approach provides a degree of independence between the visual look and the interaction feel as I'm expecting some feel differences to be desirable in the mouse vs. touch environments.

Your approach seems like an equally valid solution as well.

**Comment by memononen** [comment]
@gavinhoward thanks for keeping the PR alive, and sorry for lack of communications.

While I think the implementation is solid, I have been reluctant to merge it for 3 reasons: 1) it adds a lot of code and complexity to library (and extra files), 2) it does not handle text (I think testing glyph bboxes might be ok), 3) it's so much code that given my current free time, I cannot review it properly (i.e. there as some cases in the stroke testing I would need to dive into to understand properly, I don't like the internal header).

Solutions for pt:1 I'd much rather see `nanovg_pick.h` and `nanovg_pick.c` which can be optionally included into the project, you'd need to include `nanovg_pick.h` to use picking, and maybe call an init function too.

Solution for pt. 2: checking glyph bounding boxes should not be too hard.

Solution for pt. 3: one option would be to change the point of view of the implementation and make the picking code external dependency like `stb_truetype`. or, maybe someone could volunteer to maintain the picking part of the code?

**Comment by gavinhoward** [comment]
@memononen your reasons make complete sense.

For problem 1, I agree with that, if picking remains part of NanoVG (see below). For problem 2, I believe I can implement that. For problem 3, I am willing to maintain this, if it remains part of NanoVG.

That said, I'm not entirely sure what you mean about making the picking code an external dependency. Do you mean to make it a separate project entirely, and then those who want it could just add it as well? If so, are the internals of NanoVG exposed enough that this could be made another library built on top of NanoVG? If that is possible, I would be willing to do that.

However, if it is better as part of NanoVG, I am plenty happy to do that too, including maintaining it. 

In summary, I will also fix the three problems as soon as I can. Just let me know whether it would be best to make picking separate or not.

**Comment by lieff** [comment]
I like approach with independent code nanovg_pick.h nanovg_pick.c that can be used without NVGcontext (that is for drawing). For example in swf buttons can have invisible hit test shapes, so, why put it in drawing context? Also function that tests individual paths set will be useful (not full scene, glIsPointInFillPathNV analog).

**Comment by memononen** [comment]
@gavinhoward as for external library, I could have API like this:
```C
p = pickCreate();
pickBeginFrame(p, w, h);

pickAddPath(p, id, cmds, ncmds, scissor, strokeWidth, miterLimit, lineCap, lineJoin, strokeOrFill);
pickAddRect(p, id, ...); // for text

pickHitTest(...);

pickEndFrame(p);
pickDelete(p);
```

The point being that NanoVG would push data into the library, rather than the code pulling data from nanovg. We could design the API, so that it just happens to conveniently match the nanovg data, so we would not need to do much shuffling around. That way all the complexity is in the lib, and nanovg is just using it.

In case of you maintaining the picking code (which I warmly welcome too), I would like the implementation be more in line with the "push" API above too, so that we can get rid of the `nanovg_internal.h`. Maybe expose few things into the public header to make interop easier.

@lieff it should be possible to do that using:
```C
nvgBeginPath();
nvgRect(...);
nvgInFill(...);
```
That is, not call `nvgFill` or `nvgStroke`.

**Comment by lieff** [comment]
@memononen Yes, this is same as cairo_in_fill approach. My point 1) same as you API proposal with p = pickCreate(); (separate form NVGcontext). 2) additional utility function that operates by shape instead of pick context with many shapes. It's equivalent if I create N pick contexts and push only one shape to each context. So I can test it many times without any re-allocations. Nvidia Path Rendering have this function (glIsPointInFillPathNV), and it make things easier in swf flash player: look inside_shape https://github.com/lieff/lvg/blob/master/render/render_nvpr.c

**Comment by gavinhoward** [comment]
@memononen sorry for taking so long to reply.

The reason I took so long is because I was thinking hard about whether it would be better to make this part of NanoVG or not. And I came to the conclusion that it would be better as part of NanoVG. The biggest reason for that is that the picking code isn't really useful outside of NanoVG since it will be somewhat tied to the data structures used by NanoVG.

That said, I agree that the "push" API seems the right way to go right now. I will work on that and maintain it thereafter. And I will definitely separate the picking API into its own header.

All of that may take me a couple of weeks to do.

There seems to be a problem, however. I can't compile vanilla NanoVG on Ubuntu. It seems some required libraries are not being linked in. There was a PR to fix that (I can't find it), so I am going to open a new PR probably tomorrow (Friday for me) to fix it.

Also, I am closing this thread as things seem to be decided. Thank you.

**Comment by lieff** [comment]
@gavinhoward I can build on Ubuntu 16.04 and 17.04 without any problems. Here commands I use:
```
premake4 gmake
cd build
make
```

**Comment by gavinhoward** [comment]
@lieff I checked again, and I could not compile because I had built GLFW from source. Whoops. I guess I won't be putting in a PR.

**Comment by memononen** [comment]
@gavinhoward sounds good, go for it :) Just to clarify, I'd like to keep the current public API for picking, but change the interface between nvg and picking code so that we don't need to pull out so much internal state. Let me know if you run into any problems with that.

**Comment by gavinhoward** [comment]
@memononen will do.
